### ISSUE TITLE
SEO and Accessibility Deal Breakers

### ISSUE THREAD

**Issue by meezaan** [flarum/core#1820]
## Bug Report

**Current Behavior**
There are several issues to mention here, all related. The SEO impact of these is massive, particularly if you are going to be switching a well indexed site to Flarum.

- [ ] 1. Loading the All Discussions page with JavaScript disabled (as Google bot would see it, for instance) does not bring back The discussion titles and a post in the discussion as one would expect. Just the titles come back, which means indexing will be poor at best and those migrating to Flarum might take a hit for their homepage.
- [ ] 2. Even with JavaScript enabled, when you get to the bottom of the list and click Add More, it just loads more results. The URL needs to change so it effectively has pagination - this will really make 'all the discussions' relevant scope for a search engine bot.
- [x] 3. The 'Tags' page or a page for a single tag does not bring load anything without JavaScript. This means tag URLs will simply not get indexed by search engine bots.
- [ ] 4. With JavaScript, the single tag page does NOT change URLs for pagination when you click load more at the bottom. It must, otherwise only the first page of a tag will ever get indexed, which is bad news if your discussion thread has 5,000 posts.

It's interesting that all the content comes back without JavaScript when loading a single discussion thread. When you load with JavaScript and scroll down to load more results, the URL also changes to reflect paging. It would appear there is some inconsistency with the way the thread / discussion, tag/tags and and all discussion pages are built.

**Steps to Reproduce**
Just load the pages as mentioned above.

**Expected Behaviour**
One would expect all the content (and by this I mean semantic markup with content from the database) to be the same, with our without JavaScript - the display, of course, may not be.
Every time we click on a Load More button, the URL should paginate (otherwise the search engine bot would consider the button / link to be useless).

**Screenshots**
![All Discussions Page without JS](https://9gr6bq.bn.files.1drv.com/y4meemDx-P9j9A6pvebSbcOPBqq3aPxCZXZSKPqgbaXcIdQxeIc-yRZvk8k_EIyoSMw2yf4gJeZgDSCBYCKtEW__r80cb8FtMUzp0xmhj1tg8gm7U2xUB5e4fRYLpDMfOKtPSiWZwV4jejcj7VKK9NKvlsYXdH_8PnpCL1J9HN7BWUC0rQgKLd_FNW1fT87KYOFcWogpnGOOdCK678k61IKcA/homepage_no_js.png?psid=1)
![Tags Page without JS](https://9gtfka.bn.files.1drv.com/y4mptiazGHmvEYjY0Lfn9ZF7U_J_JIagkr-cOeWQxbFZzZMtOZNCTf4IiLSDfCH-OPyDrFOktfTbINXRq-E9FchUCAN9LSWmre-ikzqIoRAG724DwGDgGl_zYLM4uEwOuzqXmELdkpID8GD0T35QCUlEfK1OgbfP5dkjGevvGT-ZBuKWSMjVTlHirC6M9mIlgq_fqCD2Duj8W9SZoKcSy6U1g/all_tags_no_js.png?psid=1)

You can see the output of the above 2 pages without JavaScript here: https://1drv.ms/u/s!AmrVmjAdc7P2wJQvsRFe8wCUI13zRw?e=zd37Am

**Environment**
- Flarum version: 0.1.0-beta.9
- Website URL: 
- Webserver: Apache 2.4
- Hosting environment: Kubernetes
- PHP version: 7.2
- Browser: All

```
Output of "php flarum info", run this in terminal in your Flarum directory.
```

**Possible Solution**

**Additional Context**
We cannot go live with Flarum until the 2 pages render semantically correct HTML with the content.

**Comment by matteocontrini** [comment]
Some random notes:

- Google has been indexing Single Page Applications for years now, by rendering them with Chrome
- There are Flarum extensions that create a sitemap for the forum, so that search engines can have a complete and updated index of all the forum pages
- There's another extension that takes care of adding LD-JSON structured content, which also helps with indexing
- In my personal case, even without the 2 extensions mentioned above, results have been very good. Google had all the forum discussions already indexed even without a sitemap, and it was even able to extract thing like the number of posts in a discussion and the number of authors
- Flarum uses JavaScript very extensively. I think the nojs experience is there just to avoid showing a blank page, but I wouldn't expect any improvements, since not even search engines take it into account anymore

For what it's worth, my view is that there's no real reason to worry about the nojs experience, it doesn't seem to matter much in the end.

**Comment by meezaan** [comment]
Thanks for the feedback @matteocontrini - but as you rightly call it random, it doesn't really apply to the specific problems pointed out here. JSON-LD does not apply here and the extensions don't actually fix the semantic shortcomings.

For SEO, the identified issues with missing content and markup are a must. 

Accessibility issues caused by JavaScript - these are a a legal requirement for some even if they are not for all of us.

**Comment by mikejones3** [comment]
@meezaan What legal requirement are you referring too?

**Comment by meezaan** [comment]
@mikejones3 https://www.jisc.ac.uk/accessibility, for instance.

**Comment by dsevillamartin** [comment]
@meezaan Could you clarify what you mean with point flarum/framework#1 ?

As the first post's content is only shown for Sticky discussions when JavaScript is enabled, I'm not really sure what you are asking here.

**Comment by meezaan** [comment]
@datitisev Thank you for the work on this thus far. 

That's precisely it - it doesn't bring those back at all, whether they show up because of sticky discussions or the summary extension. They should at least appear in the DOM whether JS is enabled or not.

**Comment by datitisev** [comment]
@meezaan I see. That's an interesting idea. Not sure how we want to tackle that, if we do.

**Comment by meezaan** [comment]
JS should only really impact our styling or User Interface, but not the content that comes back in the DOM on the first page load - it's the same concept when viewing a single discussion.

**Comment by clarkwinkelmann** [comment]
> Loading the All Discussions page with JavaScript (as Google bot would see it, for instance) does not bring back The discussion titles and a post in the discussion as one would expect

@meezaan in your first point, do you mean *without* javascript ? Or is there an issue when the single page app is loaded inside google bot with javascript ?

The reason there's a difference between the javascript and nojs version of the page is because the Sticky extension doesn't extend both. Post previews are only implemented for javascript. We might want to add nojs support to the Sticky extension but I don't think Flarum offers any extensibility inside of the nojs page @datitisev

This is the nojs view: https://github.com/flarum/core/blob/master/views/frontend/content/index.blade.php it has no way of being extended in a clean way. We might want to improve that.

**Comment by meezaan** [comment]
@clarkwinkelmann  Yes, I meant without, sorry. Presumably we could add an if statement somewhere around https://github.com/flarum/core/blob/master/views/frontend/content/index.blade.php#L12 to check for the post preview?

**Comment by clarkwinkelmann** [comment]
It's a bit more complicated than that. Core is not aware previews are a thing. Previews are implemented by the Sticky core extension (and also by @jordanjay29 's summaries extension), so this is where the logic should go, once we make it possible for the nojs view to be extended.

**Comment by dsevillamartin** [comment]
#​1 doesn't seem to fit in core, at least not now. While more content per page does increase SEO standings, having hidden elements and loading the first post of each discussion in the discussion list probably isn't a very good idea to add to core. The rest of the issues have had PRs made to fix them.

**Comment by meezaan** [comment]
@datitisev @clarkwinkelmann OK, so 1 we can do without. I can now see all the post titles and URLs without the JavaScript, so technically it is use-able and accessible, great work.

However, when browsing all discussions or a tag, we still do not get pagination appearing in the URL. These are #2 and #4 in the list. If there is a PR for these issues, can you link it to this issue, please.

Thanks.

**Comment by meezaan** [comment]
@clarkwinkelmann Thank you. I think you've captured most of it but SEO is also dependent on links to the website from third parties. Currently, I don't have a way to link the 3rd page for a particular tag, along with the correct ```rel=next``` meta tag and a canonical url tag we would need to allow changing the URL via load more to enable that kind of linking too. I think the rest here makes sense.

**Comment by clarkwinkelmann** [comment]
Is there a rationale for deep-linking to paginated discussion results from third-party websites ? The linked content won't stay the same.

Same for search engines, it's great if they can use the pagination to discover content, but by the time a user clicks the paginated link, it's possible what they are looking was already pushed multiple pages away. So the sitemap makes more sense to me than letting search engines browsing paginated results.

The only useful reason I see for user-visible pagination in the url, is if you refresh the page so you end up at about the same place.

**Comment by dsevillamartin** [comment]
@meezaan That'd be flarum/framework#1829 which adds pts 2 & 4.

**Comment by meezaan** [comment]
@datitisev Thank you.
@clarkwinkelmann You're right - paginated tags or All Discussions is not useful as it would change (the default sort order is last_updated_at), but a single paginated discussion is useful. In the latter case, if you want to reference a particular page in a large discussion thread with relevant posts, you would need paginated URLs.

**Comment by luceos** [comment]
So what's the status on this, are all items except no. 1 tackled?

**Comment by meezaan** [comment]
1 we can do without. In theory, we can close this issue.

We should probably track stuff similar to what is mention in https://github.com/flarum/core/issues/1820#issuecomment-520911064 somewhere.

**Comment by selfthinker** [comment]
Just to clarify the potential accessibility issues...

If something doesn't work without JavasScript that's not strictly speaking an accessibility issue. It would definitely not fall under the regulations that @meezaan linked to as it's not covered by WCAG on which the regulations are based on.
It does not affect someone with an impairment more than it would someone without an impairment. You could make an argument, though, that people with impairments often have older technology. And having older technology has an impact on JavaScript.

Something not working without JavaScript is more about resilience, robustness and usability.
People usually talk about noJS like it's only affecting people who intentionally switch JavaScript off. But that is not the case. That is only the minority.

In 2013 [GDS found that 1.1% of visits were not getting JS](https://gds.blog.gov.uk/2013/10/21/how-many-people-are-missing-out-on-javascript-enhancement/), but only 0.2% of those because they have JS disabled or their browser doesn't support it. 0.9% don't have JS for other reasons.
(Some people say these numbers would be much lower nowadays. My personal theory is that they would be much higher. But we don't have any newer numbers, unfortunately.)

That blog post wrongly says it's 1.1% of all **users**, but it was 1.1% of all **visits**. Over time it can affect 100% of users. [See what that distinction means and why it matters.](http://www.kryogenix.org/code/browser/why-availability/)

If people don't switch JS off and their browser supports it, there are lots of [other reasons for not getting the JS](http://kryogenix.org/code/browser/everyonehasjs.html).

So, there are lots of good reasons for making the noJS experience better, but accessibility is not really one of them.

**Comment by meezaan** [comment]
@selfthinker I think the intent here was, if I may summarize, that content should be visible with our without JS, and should be navigable (one way or another) for someone who does not load with JS.

**Comment by selfthinker** [comment]
@meezaan, I'm confused. Was there anything in my comment that made you think I was talking about something else?

**Comment by meezaan** [comment]
@selfthinker Not at all - much like the way I had initially stated, there is a lot of information - but what we are aiming for is just one line amongst the many and can get lost - so stating exactly what we are after always helps.

**Comment by askvortsov1** [comment]
Splitting into https://github.com/flarum/core/issues/2537, https://github.com/flarum/core/issues/2173
